### ISSUE TITLE
LPE Mavlink Message 

### ISSUE THREAD

**Issue by tuofeichen** [PX4/Firmware#3901]
Hi @mhkabir,
I followed the instruction here to upload the LPE version of the firmware onto the pixhawk: https://github.com/PX4/Firmware/pull/3209, and disabled INAV for the system to use LPE.

I checked QGC that indeed the LPE is enabled and all the parameters are available to tune. Then I tried to use the mavlink inspector to look at the `LOCAL_POSITION_NED` topic. But the Z value is still quite off. (always some random value below 1)

I'm using the px4flow built in sonar and by looking at the `DISTANCE_SENSOR`  topic, it does output the right value. 

I'm wondering if I need to tune any paramater for the LPE to add more weight on the sonar, or if LPE simply doesn't publish the local estimate to LOCAL_POSITION_NED?

Thank you!

Best
Francis

**Comment by jgoppert** [comment]
The latest lpe code is here: https://github.com/PX4/Firmware/pull/3748

Is the light green? If not, you need to check your optical flow camera. Look at the px4flow image in qgc and see if it is in focus at the height you want to fly, next, hold it at 1 m for about 5 seconds, the light should turn green and start publishing local position.

**Comment by tuofeichen** [comment]
I figured most of it out by looking at the code and yes indeed I hold it until the light turned green and it start publishing local position but it's still kinda off. 

Ok I will try upload the latest code. Should I directly upload the master or your lpe_improvements branch? 

Thank you!

Best
Francis

**Comment by tuofeichen** [comment]
also, just tried it out. Seems like its working perfectly! 

Thank you so much. Saves me a lot of trouble hacking into the firmware!

**Comment by jgoppert** [comment]
You should checkout the lpe_improvements branch and then build with:

```bash
make px4fmu-v2_lpe upload
```

**Comment by tuofeichen** [comment]
Gotcha, did that and the z estimate is quite well now. I'm wondering since I'm not mounting the flow module at the center of the bottom, is there a parameter I can change, that's similar to the INAV_FLOW_DIST_X and INAV_FLOW_DIST_Y? 


Thanks

Best
Francis

**Comment by jgoppert** [comment]
There is a flow z offset. It shouldn't matter for xy offset if it is less
than 10 cm or so from the center. The rotation effects at that distance are
negligible.

**Comment by tuofeichen** [comment]
Ok sounds good. Just to clarify, if I were able to do achieve stable hovering in position control mode using this setup, I should also be able to do offboard local_position_setpoint (using mavros) with this as well, is that correct?  

Thanks 

Best
Francis

**Comment by jgoppert** [comment]
@tuofeichen Yes, I had stable position mode. I think it was a bit jumpy. I had the max vel. set at 5 m/s so I think you will want to bump this down to 1 m/s or so. Yes, the local position setpoints should work but I haven't tried it yet.

**Comment by tuofeichen** [comment]
sounds good, just out of curiosity, I use the masters's Flow firmware, it should be fine right? or should I also flash my flow module with the flow firmware in your repo? 

Best
Francis

**Comment by jgoppert** [comment]
@tuofeichen You can use the firmware for PX4Flow on the master branch. Mine just adds some better debug output.

**Comment by tuofeichen** [comment]
Hi @jgoppert ,
I was using your LPE improvements around that time with only flow and sonar (no GPS) and the Pixhawk turned green just fine in LPE. I tried to handhold and test the position estimate walking around a square and get the plot below (With blue as the raw flow estimate, which closes the loop really well and red as the lpe estimate). The raw flow estimate is really accurate and closes the loop, but the LPE seems to have some trouble when the drone is turning. I'm looking at your code to figure out why, but if anything comes up right way let me know.

![Image of Flow vs LPE] (http://i63.tinypic.com/30jpef7.jpg)

![image of x,y, z and quality] (http://i64.tinypic.com/17a683.jpg)

Recently, I see you merge it to master and uploaded the new Firmware onto the Pixhawk (with same setup. flow+sonar, no GPS). However once I did that, even the flow quality is 255, the LED never turns green any more and the local position estimate never updates. Anything changed with that respect I should be aware of?

**Comment by jgoppert** [comment]
@tuofeichen Thanks for the analysis. The LPE estimator keeps the optical flow integrated position as a separate state. As long as the flow is high quality as you have plotted, the correction should keep it close to the pure optical flow result. It isn't clear what version of the code you are running, so if you could try this branch: https://github.com/PX4/Firmware/pull/4036 it would help. I just flew that branch with optical flow and lidar and it worked very well.

**Comment by tuofeichen** [comment]
@jgoppert , I tried to upload the lpe_default branch, same thing occurs to me that the LED never turns green, despite a very good flow measurement. Let me know if there's anything I should be aware of. The version I test out is your lpe-improvements branch, but again, that's been a while ago and its hard to trace back.

**Comment by jgoppert** [comment]
Do you receive any messages from lpe in qgroundcontrol?

**Comment by tuofeichen** [comment]
Yes, I received all the message about [lpe] flow init and so forth, I
edited my comment and it seemed to me that LPE is not detecting the sonar
for some reason

**Comment by jgoppert** [comment]
Have you tried upgrading the firmware on the flow? Also, you can plug it into qgroundcontrol and see the optical flow and sonar data directly. That will let you know if your device is functioning properly.

**Comment by jgoppert** [comment]
@tuofeichen Also, posting the exact messages you are getting will be helpful for me to debug it if it isn't the device.

**Comment by tuofeichen** [comment]
I get all the flow data stand alone just fine, including the accurate
reading of sonar. In fact the mavlink message optical-flow-raw is
outputting the right thing as well. Just the Lpe is not recognizing the
sonar, and constantly output a no distance for flow error.

This not only happens for the Lpe-default branch but also for the master
branch.

The old Lpe-improvements branch that I used to generatw the plot was still
able to recognize the sonar. I wonder what changed since then

**Comment by tuofeichen** [comment]
Yeah sure will compile the message that I get with some screenshots. Thanks
!

**Comment by jgoppert** [comment]
@tuofeichen I'm trying on my end right now with flow on my 3dr iris. Did you try upgrading your firmware version for the px4flow to the latest version on master?

**Comment by tuofeichen** [comment]
Hi I flashed my flow firmware quite recently. I will do it again just to
make sure. I'm on it right now as well. Just give me a sec!

**Comment by jgoppert** [comment]
@tuofeichen I can confirm the bug. Fixing it now.

**Comment by tuofeichen** [comment]
Yeah ok cool. It's kinda weird. Wouldn't expect something significant. I
flash the flow and it still didn't fix the problem. Lemme know what's wrong
with it!

**Comment by tuofeichen** [comment]
@jgoppert , I'm a little confused at this part https://github.com/jgoppert/Firmware/blob/lpe_default/src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp#L253 .  Sorry I'm really new to this code base. I tried to publish some mavlink message like this  for debugging right after that line.

 ` mavlink_log_info(_mavlink_fd, "round %d, reading sensor type %d", i,	_distance_subs[i]->get().type);` 

 But I got this as the output:
`
 [INFO] [1433864930.019561767]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.071136474]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.122894287]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.174835205]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.224517822]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.276580810]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.329284667]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.383148193]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.433990478]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.484893798]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.536651611]: FCU: round 3, reading sensor type 0
[ INFO] [1433864930.588531494]: FCU: round 3, reading sensor type 0
`

shouldn't the message at least loop through round 0 to round 3? again, really new to this code base and don't know what's going on here, maybe you can explain it a bit?

Thanks. Also let me know what's the issue with the sonar detection.

Thanks
Francis

**Comment by jgoppert** [comment]
@tuofeichen Is seems that distance data from flow was turned off recently. So that explains our issues.

https://github.com/PX4/Firmware/commit/47f79a66d146e9be8a914aa5918aaef934eb618d

I can fix it, the flow message has distance embedded in it.

**Comment by tuofeichen** [comment]
Cool sounds good. That explains things. Let me know when you're ready so
that I can test it out!

**Comment by jgoppert** [comment]
@tuofeichen Works now. Try lpe_default branch. I just pushed a commit that adds back publication within LPE of the flow distance.

**Comment by tuofeichen** [comment]
Sounds good. I'm just back from the lab. Will test it out tomorrow and let
you know how it goes.

Best
Francis

**Comment by jgoppert** [comment]
@tuofeichen Note that the green light only turns on when global position is valid. This only occurs if you get at least one gps message to initialize the local reference frame. But you don't need global position to do position hold. Just tested it.

**Comment by tuofeichen** [comment]
Hmm, ok, this might be a little tricky. I don't have a GPS with me right
now. Can I fake a global position message ?

**Comment by jgoppert** [comment]
@tuofeichen No need for GPS. I just did a test flight with my GPS unplugged, it works.

**Comment by tuofeichen** [comment]
Sounds good, I will test it out tmr and let you know.

**Comment by jgoppert** [comment]
Having the green light is really useful, so I just added a parameter for the initial lat/lon of the vehicle.

**Comment by tuofeichen** [comment]
@jgoppert , tested out the lpe_default branch, it works without GPS and light turns green. Did the same test, would say it definitely have a better performance. But it still has some trouble closing the loop. 

![Image of  XY trajectory] (https://i.imgsafe.org/1713ba1.jpg)

![Image of flow lpe and quality] (https://i.imgsafe.org/a95b577.jpg)

I looked into the orientation estimate of the LPE and it seems fine to me. Also tried to disable LPE_INTEGRATE, still doesn't get better. 

Another thing I notice is, the LPE's z estimate is not very stable. When it's sitting on a surface, meaning the flow sonar is timeout, the Z estimate is still 1.5m or so. I suppose this gets from the baro reading and I can disable it quite easily. But Idk if this will be any interest to you. I'm flying indoor btw, so the baro is not very reliable.

**Comment by jgoppert** [comment]
I don't see how they could be that different. How are you getting the
integrated flow plot, where is the data coming from? Try setting the flow
standard deviation down, then xy should be almost exactly the same as
integrated flow.

**Comment by jgoppert** [comment]
Also, flow depends on resolving the body flow to nav frame,  so if there
are any computers etc that you are walking past, that would change the
compass heading and cause a distorted square like you are showing.

**Comment by tuofeichen** [comment]
1. I got the integrated flow plot from the `flow_comp_m_x` and `flow_comp_m_y` of the `OPTICAL_FLOW` message. And the stdev of flow is set to minimum 0.01. Since its updated at 15 hz, I times 1/15s to integrate it.

2. I'm aware of the frame transform. That's why I was looking at the orientation estimate of the LPE, and it seems to remain stable close to x y z = 0. w = 1. (I'm not rotating the drone at all)

**Comment by jgoppert** [comment]
Well try replacing flow integration in lpe with your method,  see if it
outperforms the more complicated method.

**Comment by tuofeichen** [comment]
alright will let you know when I get on that

**Comment by jgoppert** [comment]
@tuofeichen you could also set the flow gyro high pass filter very high to see if that gives you better results

**Comment by jgoppert** [comment]
Very high will eliminate gyro compensation,  and very low will use gyro including dc offset, which is a bad thing.

**Comment by tuofeichen** [comment]
@jgoppert , yeah I'm looking into the gyro compensation as well. Since the raw optical flow is not using the gyro compensation. Can you explain this line to me? I'm a little confused by what this delta_b stands for. 

https://github.com/jgoppert/Firmware/blob/lpe_default/src/modules/local_position_estimator/BlockLocalPositionEstimator.cpp#L1166

**Comment by tuofeichen** [comment]
If that's the rotation of the body, why are we subtracting optical flow rad from gyro rad?

**Comment by jgoppert** [comment]
@tuofeichen The logic is that optical flow observed by the camera is due to two components:

rotation of the camera and translation of the camera. We want to isolate the translation of the camera, by subtracting the flow due to the rotation of the px4flow, which is measure by the gyro

But we don't want any steady state bias in the gyro driving off the flow integral, so we just take the high frequency components of the gyro signal (hence the high pass filter)

**Comment by tuofeichen** [comment]
I see, makes sense. So that delta_b is the translation of the flow?

**Comment by jgoppert** [comment]
Yes, in the body frame, which is then rotated to delta_n (navigation frame NED)

**Comment by tuofeichen** [comment]
Right right, delta_n is the global frame, ok cool

**Comment by tuofeichen** [comment]
@jgoppert, ok so I was able to narrow down the issue to the attitude estimate. I log the integrated flow using the LPE's method, while only integrating the body frame and get a very good estimate of the loop (and pretty much the same as my old method). See below. Thus the gyro compensation on the Flow is not causing the problem. 
![Image of Flow] (https://i.imgsafe.org/22d42e4.jpg)

But after the transformation, it becomes distorted, especially the bottom right corner. I logged the quarternion and convert to rotation matrix and multiply by my logged local frame estimate and got the following: 

![Image of Flow] (https://i.imgsafe.org/0036bb4.jpg)

I'm going to look into the attitude_q_estimator and see what happens. I checked that corner when I walk by, doesn't seem to have a big magnetic field. but I will double check. Anyway, I think I'm pretty close to figure it out. Also it seems like the problem's not on the LPE side. 

Anyway, thanks for walking me through this, I will let you know what I figure out. 

Best
Francis
